cudatoolkit: 10.2 -> 11.5

SomeoneSerge opened the pull request:
also downgrade cudatoolkit_11: 11.5 -> 11.4 until pytorch 1.11 passes staging

###### Description of changes

As discussed in: https://github.com/NixOS/nixpkgs/pull/166784/#issuecomment-1086762659
The motivation is to build as many packages as possible with the "defaults": sharing cuda, cudnn, magma, and nccl instances and simplifying overlays.
This is a major bump for cuda version, but a small step in pruning this garden:)


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] [x86_64-linux](https://hercules-ci.com/github/SomeoneSerge/nixpkgs-unfree/jobs/169)
- [ ] (inprogress) Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)

CC @NixOS/cuda-maintainers

SomeoneSerge committed:
cudatoolkit: 10.2 -> 11.4

also downgrade cudatoolkit_11: 11.5 -> 11.4

samuela reviewed (approved):
LGTM once all the packages build I say we merge!

SomeoneSerge committed:
python3Packages.mxnet: pass explicit cuda gencodes

SomeoneSerge commented on pkgs/applications/science/math/mxnet/default.nix:
```diff
@@ -2,6 +2,7 @@
 , opencv3, gtest, blas, gomp, llvmPackages, perl
 , cudaSupport ? config.cudaSupport or false, cudatoolkit, nvidia_x11
 , cudnnSupport ? cudaSupport, cudnn
+, cudaCapabilities ? [ "3.7" "5.0" "6.0" "7.0" "7.5" "8.0" "8.6" ]
```
This is so as to remove `sm_30` from mxnet's default gencode list. The `sm_30` has been deprecated in cuda 11. I'm not inventing any naming schemes here and just using the format that the package accepts - we'll need to rework this when we introduce `config.cudaCapabilities` or anything similar

SomeoneSerge commented on pkgs/applications/science/math/mxnet/default.nix:
```diff
@@ -2,6 +2,7 @@
 , opencv3, gtest, blas, gomp, llvmPackages, perl
 , cudaSupport ? config.cudaSupport or false, cudatoolkit, nvidia_x11
```
Hmm, maybe I should get rid of this `nvidia_x11` while I'm at it? But this would demand some testing

samuela replied:
If you're up for it, but it's also prob not a blocker for this PR

SomeoneSerge replied:
I guess it's safer to keep focus

SomeoneSerge commented:
```
34 packages updated:
colmap cudnn (10.2-cudnn-8.3.2 → 11.4-cudnn-8.3.2) cutensor (10.2-cutensor-1.3.1.3 → 11.4-cutensor-1.3.1.3) cudaPackages.cudatoolkit_11 (11.5.0 → 11.4
.2) ethminer ethminer forge gpu-burn-unstable gpu-screen-recorder gpu-screen-recorder-gtk gromacs gwe katago librealsense librealsense librealsense ma
gma nccl (2.7.8-1-cuda-10.2 → 2.7.8-1-cuda-11.4) nccl_cudatoolkit_11 (2.7.8-1-cuda-11.5 → 2.7.8-1-cuda-11.4) nvtop python3.10-cupy python3.10-jaxlib p
ython3.10-pycuda python3.10-pynvml python3.10-Theano python3.9-cupy python3.9-jaxlib python3.9-pycuda python3.9-pynvml python3.9-pytorch python3.9-ten
sorflow-gpu python3.9-Theano truecrack xpra
```


Failing:

- [ ] /nix/store/k7z28nmwa7q6azykhsmrciihwa1v0xxi-gpu-screen-recorder-1.0.0.drv
  `<cuda.h>`
- [ ] /nix/store/6dnifwyqhx171v6xw9ysws1b5c9fgfkl-ethminer-0.19.0.drv
   `> nvcc fatal   : Unsupported gpu architecture 'compute_30'`
- [ ] `/nix/store/0xqz25b6ckli13xzz7gf2qlhycskvqi1-truecrack-3.6.drv`
  hardcodes `-gencode arch=compute_20,code=sm_20 -gencode arch=compute_30,code=sm_30 -gencode ar
ch=compute_35,code=sm_35`
- TBC

SomeoneSerge committed:
ethminer: fix cuda11 support

cuda11 removes support for compute_30
cosmetic changes: move postPatch and patches closer to src

SomeoneSerge committed:
gpu-screen-recorder{,-gtk}: only build with cuda10

SomeoneSerge committed:
truecrack: doesn't build with cudaSupport

SomeoneSerge commented on pkgs/tools/security/truecrack/default.nix:
```diff
@@ -33,6 +33,7 @@ gccStdenv.mkDerivation rec {
   meta = with lib; {
     description = "TrueCrack is a brute-force password cracker for TrueCrypt volumes. It works on Linux and it is optimized for Nvidia Cuda technology.";
     homepage = "https://gitlab.com/kalilinux/packages/truecrack";
+    broken = cudaSupport;
```
CC @ethancedwards8 I couldn't build `truecrack-cuda` neither in `nixpkgs-unstable` nor in `master`, neither with cuda10 or cuda11

ethancedwards8 replied:
Just saw this. Thanks for the information. Sorry for not responding earlier. Currently, I'm no longer using NixOS full time, nor am I using Truecrack, so I'm afraid I'm not willing to fix it. I believe building without cuda support works, right? If so, I don't see much of a reason to worry. If someone was using cuda support, I imagine someone would have created an issue.

SomeoneSerge replied:
👍🏻 @ethancedwards8  No problem, I'm only trying to keep track of what works and what doesn't in nixpkgs when it comes to acceleation.

You also can remove yourself from the maintainer list for the package to stop receiving notifications about it

SomeoneSerge commented on pkgs/top-level/all-packages.nix:
```diff
@@ -26411,9 +26411,15 @@ with pkgs;
-  gpu-screen-recorder = callPackage ../applications/video/gpu-screen-recorder { };
+  gpu-screen-recorder = callPackage ../applications/video/gpu-screen-recorder {
+    # rm me as soon as this package gains the support for cuda 11
+    cudatoolkit = cudatoolkit_10;
```
I couldn't find means to remove `sm_30` from the gencode list, so building with cuda10 for now

CC @babbaj

SomeoneSerge commented on pkgs/tools/misc/ethminer/default.nix:
```diff
@@ -31,6 +32,21 @@ stdenv.mkDerivation rec {
       fetchSubmodules = true;
     };
 
+  patches = [
```
This whole chunk only really adds `fetchpatch` with a fix for cuda 11. The rest is moving `postPatch` and `patches` from the bottom closer to `src`

SomeoneSerge commented:
I think these were the only failures 🤔 
I'm re-running the build: https://hercules-ci.com/github/SomeoneSerge/nixpkgs-unfree/jobs/183
And I'll run nixpkgs-review again

SomeoneSerge committed:
python310Packages.mxnet: mark cudaSupport broken

SomeoneSerge commented on pkgs/applications/science/math/mxnet/default.nix:
```diff
@@ -60,6 +62,12 @@ stdenv.mkDerivation rec {
+  # used to mark cudaSupport in python310Packages.mxnet as broken;
+  # other attributes exposed for consistency
+  passthru = {
+    inherit cudaSupport cudnnSupport cudatoolkit cudnn;
```
CC @abbradar 

cmake fails to configure in python 3.10 when enabling cuda and python

SomeoneSerge committed:
opensubdiv: disable opencl when building cuda

SomeoneSerge commented on pkgs/development/libraries/opensubdiv/default.nix:
```diff
@@ -37,15 +41,18 @@ stdenv.mkDerivation rec {
+      "-DOSD_CUDA_NVCC_FLAGS=--gpu-architecture=${cudaArch}"
       "-DCUDA_HOST_COMPILER=${cudatoolkit.cc}/bin/cc"
+    ] ++ lib.optionals (!openclSupport) [
+      "-DNO_OPENCL=1"
```
This appears to unbreak cuda-blender with less effort than https://github.com/NixOS/nixpkgs/pull/167259

SomeoneSerge committed:
python3Packages.chainer: mark cudaSupport broken

SomeoneSerge committed:
python310Packages.caffe: mark as broken

SomeoneSerge commented:
tldr: there are just two failures remaining, cholmod-extra and mxnet; both are unrelated to the PR

https://hercules-ci.com/github/SomeoneSerge/nixpkgs-unfree/jobs/189

samuela commented:
Will merging this cause us merge conflict headaches with https://github.com/NixOS/nixpkgs/pull/167016?

samuela commented:
If not, then I'm cool with merging!

SomeoneSerge commented:
Pytorch 1.11 has been merged into master, so I'm reworking this with 11.6!

SomeoneSerge committed:
cudatoolkit_11: 11.4 -> 11.6

SomeoneSerge commented:
> Will merging this cause us merge conflict headaches with https://github.com/NixOS/nixpkgs/pull/167016?

I've tried rebasing the redist PR on top of this one: I just had to drop the parameters in the call to pytorch and some minor stuff in mxnet. Waiting for the checks to run. Here's the branch: https://github.com/SomeoneSerge/nixpkgs/tree/cudatoolkit-redist-on-11

samuela commented:
sounds good, I think it makes sense to merge this one first since it will help us to set `cudaPackages = cudaPackages_11_6` when merging https://github.com/NixOS/nixpkgs/pull/167016.

SomeoneSerge committed:
cudnn 8.3.2: unbreak for cuda 11.6

It has been added to the official support matrix:
https://docs.nvidia.com/deeplearning/cudnn/support-matrix/#cudnn-cuda-hardware-versions

SomeoneSerge committed:
cudatoolkit: 11.6 -> 11.5

SomeoneSerge commented:
I was wrong about cudnn 8.3.2 officially supporting cuda 11.6 (although de facto it does), so I'm instead updating to 11.5: it's support both by current cudnn and by the newly merged pytorch

samuela commented:
> I was wrong about cudnn 8.3.2 officially supporting cuda 11.6 (although de facto it does), so I'm instead updating to 11.5: it's support both by current cudnn and by the newly merged pytorch

That's correct. Although I think cuDNN 8.3.3 has been released recently with support for CUDA 11.6.

SomeoneSerge commented:
Yep, and it seems that adding a new cudnn version is so much easier with the interface that https://github.com/NixOS/nixpkgs/pull/167016 offers :)

FRidh commented:
> I've tried rebasing the redist PR on top of this one: I just had to drop the parameters in the call to pytorch and some minor stuff in mxnet. And a few more changes related to staging merging into master. Waiting for the checks to run. Here's the branch: https://github.com/SomeoneSerge/nixpkgs/tree/cudatoolkit-redist-on-11

Good! Then I won't touch my PR until this is in.

SomeoneSerge commented:
All builds (except one, with an old overlay) complete: https://hercules-ci.com/github/SomeoneSerge/nixpkgs-unfree/jobs/201
